pkg/operator/ceph/cluster/osd/replace.go

```diff
@@ -0,0 +1,208 @@
+func (q *ReplaceQueue) Pop() (osd *OSDReplaceInfo, ok bool) {
+	if q.Len() == 0 {
+		return nil, false
+	}
+
+	osd = &q.Q[0]
+	q.Q = q.Q[1:]
+	return osd, true
+}
+
+// OSDReplaceConfig contains the status of the OSD migration process
+type OSDReplaceConfig struct {
+	// OSDs that require migration
+	OSDs           *ReplaceQueue   `json:"osdList,omitempty"`
+	IsPgHealthy    bool            `json:"isPgHealthy,omitempty"`
+	OSDInProgress  *OSDReplaceInfo `json:"osdInProgress,omitempty"`
+	StartTimeStamp time.Time       `json:"startTimeStamp,omitempty"`
+}
+
+func newOSDReplaceConfig() *OSDReplaceConfig {
+	return &OSDReplaceConfig{
+		StartTimeStamp: time.Now(),
+	}
+}
+
+func (o *OSDReplaceConfig) createOrUpdate(context *clusterd.Context, clusterInfo *cephclient.ClusterInfo) error {
+	configStr, err := o.string()
+	if err != nil {
+		return errors.Wrapf(err, "failed to convert osd replace config to string")
+	}
+
+	existingCM, err := context.Clientset.CoreV1().ConfigMaps(clusterInfo.Namespace).Get(clusterInfo.Context, OSDReplaceConfigName, metav1.GetOptions{})
+	if err != nil {
+		if !kerrors.IsNotFound(err) {
+			return errors.Wrapf(err, "failed to get existing osd config map")
+		}
+
+		newConfigMap := &v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      OSDReplaceConfigName,
+				Namespace: clusterInfo.Namespace,
+			},
+			Data: map[string]string{
+				OSDReplaceConfigKey: configStr,
+			},
+		}
+
+		// Set controller reference only when creating the configMap for the first time
+		err = clusterInfo.OwnerInfo.SetControllerReference(newConfigMap)
+		if err != nil {
+			return errors.Wrapf(err, "failed to set owner reference on configMap %q", newConfigMap.Name)
+		}
+
+		_, err = context.Clientset.CoreV1().ConfigMaps(clusterInfo.Namespace).Create(clusterInfo.Context, newConfigMap, metav1.CreateOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "failed to create osd replace configMap %q", newConfigMap.Name)
+		}
+
+		return nil
+	}
+
+	// Update existing configMap
+	// TODO: update only if data has changed
+	existingCM.Data[OSDReplaceConfigKey] = configStr
+	if _, err := context.Clientset.CoreV1().ConfigMaps(clusterInfo.Namespace).Update(clusterInfo.Context, existingCM, metav1.UpdateOptions{}); err != nil {
+		return errors.Wrapf(err, "failed to update existing OSD replace config map %q", existingCM.Name)
+	}
+
+	return nil
+}
+
+func (o *OSDReplaceConfig) delete(context *clusterd.Context, clusterInfo *cephclient.ClusterInfo) error {
+	err := context.Clientset.CoreV1().ConfigMaps(clusterInfo.Namespace).Delete(clusterInfo.Context, OSDReplaceConfigName, metav1.DeleteOptions{})
+	if err != nil && !kerrors.IsNotFound(err) {
+		return errors.Wrapf(err, "failed to delete the osd replace status config map %q", OSDReplaceConfigName)
+	}
+
+	return nil
+}
+
+func (o *OSDReplaceConfig) string() (string, error) {
+	configInBytes, err := json.Marshal(o)
```

Comment:
If it's just OSD IDs, it will be simple for a comma-separated list as a string, instead of requiring json.

Reply:
lets discuss this more in the huddle today.

Reply:
?

Reply:
Per discussion, this is still used for the osd in progress